src/data/gaussian_int.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2019 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Chris Hughes
+
+The gaussian integers ℤ[i].
+-/
+
+import number_theory.pell data.complex.basic algebra.euclidean_domain algebra.associated
+open zsqrtd complex
+
+@[reducible] def gaussian_int : Type := zsqrtd (-1)
+
+local notation `ℤ[i]` := gaussian_int
+
+namespace gaussian_int
+
+instance : has_repr ℤ[i] := ⟨λ x, "⟨" ++ repr x.re ++ ", " ++ repr x.im ++ "⟩"⟩
+
+instance : comm_ring ℤ[i] := zsqrtd.comm_ring
+
+def norm (x : ℤ[i]) : ℕ := x.re.nat_abs * x.re.nat_abs + x.im.nat_abs * x.im.nat_abs
```

Comment:
I suggest making this a definition for `zsqrtd`, and turning this into a lemma `norm_eq`. There are already statements like `is_pell_norm` that have norm in their name, but not in the statement. That could then also be "fixed". Later we will want to generalise the norm to arbitrary number fields.

Reply:
So the norm on `zsqrtd` is `| ⟨x, y⟩ := x*x - d*y*y`? This isn't always a natural number, you need `d` to be negative for that. I guess I could define it on `zsqrtd` where `d` is negative. Is there any nice way of generalising the Euclidean domain result?

Reply:
I also think `zsqrtd` should probably be moved outside `pell`

Reply:
note that norms in thealgebraic sense are often not positive; it's not the same sense of `norm` as `normed_ring`

Reply:
Good point. Yes, norms of quadratic integers can be negative, so the norm function should be a function to the integers. It is nontrivial to determine when the corresponding quadratic extension is a Euclidean domain: https://en.wikipedia.org/wiki/Quadratic_integer.
I will leave it to you to decide if it is worth defining integer-valued norms on `zsqrtd` now, or whether it is best left to a later day. In the latter case, just let me know, and I will merge.